File: samples/TeamsSDK/Archived/msgext-action-quickstart/python/bots/bot_activity_handler.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from http import HTTPStatus

from botbuilder.core import CardFactory, InvokeResponse, TurnContext
from botbuilder.core.teams import TeamsActivityHandler
from botbuilder.schema import HeroCard
from botbuilder.schema._models_py3 import CardImage
from botbuilder.schema.teams import (
    MessagingExtensionAction,
    MessagingExtensionActionResponse,
    MessagingExtensionAttachment,
    MessagingExtensionResult,
)

SHARED_IMAGE_URL = (
    "https://encrypted-tbn0.gstatic.com/images?"
    "q=tbn:ANd9GcQtB3AwMUeNoq4gUBGe6Ocj8kyh3bXa9ZbV7u1fVKQoyKFHdkqU"
)


class BotActivityHandler(TeamsActivityHandler):
    """Handles messaging extension action commands."""

    async def on_invoke_activity(self, turn_context: TurnContext) -> InvokeResponse:
        # Handle composeExtension/submitAction here directly so the response
        # model is returned to the CloudAdapter without being pre-serialized
        # by TeamsActivityHandler._create_invoke_response (which would lead to
        # double-serialization in CloudAdapter.process).
        if turn_context.activity.name == "composeExtension/submitAction":
            action = MessagingExtensionAction().deserialize(
                turn_context.activity.value
            )
            response = await self.on_teams_messaging_extension_submit_action(
                turn_context, action
            )
            return InvokeResponse(status=int(HTTPStatus.OK), body=response)

        return await super().on_invoke_activity(turn_context)

    async def on_teams_messaging_extension_submit_action(
        self,
        turn_context: TurnContext,
        action: MessagingExtensionAction,
    ) -> MessagingExtensionActionResponse:
        if action.command_id == "createCard":
            return self._create_card_command(action)
        if action.command_id == "shareMessage":
            return self._share_message_command(action)
        return MessagingExtensionActionResponse()

    def _create_card_command(
        self, action: MessagingExtensionAction
    ) -> MessagingExtensionActionResponse:
        data = action.data or {}
        card = CardFactory.hero_card(
            HeroCard(
                title=data.get("title", ""),
                subtitle=data.get("subTitle", ""),
                text=data.get("text", ""),
            )
        )
        return self._build_response(card)

    def _share_message_command(
        self, action: MessagingExtensionAction
    ) -> MessagingExtensionActionResponse:
        message_payload = action.message_payload

        user_name = "unknown"
        if (
            message_payload
            and message_payload.from_property
            and message_payload.from_property.user
        ):
            user_name = (
                message_payload.from_property.user.display_name or user_name
            )

        include_image = (action.data or {}).get("includeImage") == "true"
        images = [CardImage(url=SHARED_IMAGE_URL)] if include_image else []

        attachments_count = (
            len(message_payload.attachments)
            if message_payload and message_payload.attachments
            else 0
        )
        body_content = ""
        if message_payload and message_payload.body:
            body_content = message_payload.body.content or ""

        card = CardFactory.hero_card(
            HeroCard(
                title=f"{user_name} originally sent this message:",
                text=body_content,
                subtitle=f"({attachments_count} attachments not included)",
                images=images,
            )
        )
        return self._build_response(card)

    @staticmethod
    def _build_response(card) -> MessagingExtensionActionResponse:
        attachment = MessagingExtensionAttachment(
            content=card.content,
            content_type=card.content_type,
            preview=card,
        )
        return MessagingExtensionActionResponse(
            compose_extension=MessagingExtensionResult(
                type="result",
                attachment_layout="list",
                attachments=[attachment],
            )
        )

File: samples/TeamsSDK/Archived/msgext-action-quickstart/js/botActivityHandler.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { TeamsActivityHandler, CardFactory } = require('botbuilder');

class BotActivityHandler extends TeamsActivityHandler {
    /*
        Building a messaging extension action command is a two-step process:
        (1) Define how the messaging extension looks and is invoked in the client
            (via the app manifest / Developer Portal).
            Learn more: https://aka.ms/teams-me-design-action.
        (2) Define how the bot service responds to incoming action commands.
            Learn more: https://aka.ms/teams-me-respond-action.
    */

    // Invoked when the service receives an incoming action submit.
    handleTeamsMessagingExtensionSubmitAction(context, action) {
        switch (action.commandId) {
            case 'createCard':
                return createCardCommand(action);
            case 'shareMessage':
                return shareMessageCommand(action);
            default:
                throw new Error('NotImplemented');
        }
    }

    // Messaging Extension - Link Unfurling.
    handleTeamsAppBasedLinkQuery(context, query) {
        const attachment = CardFactory.thumbnailCard(
            'Thumbnail Card',
            query.url,
            ['https://raw.githubusercontent.com/microsoft/botframework-sdk/master/icon.png']
        );

        return {
            composeExtension: {
                attachmentLayout: 'list',
                type: 'result',
                attachments: [attachment]
            }
        };
    }
}

// 'Create Card' compose-box command - builds a hero card from user input.
function createCardCommand(action) {
    const { title, subTitle, text } = action.data;
    const heroCard = CardFactory.heroCard(title, text);
    heroCard.content.subtitle = subTitle;

    const attachment = {
        contentType: heroCard.contentType,
        content: heroCard.content,
        preview: heroCard
    };

    return {
        composeExtension: {
            type: 'result',
            attachmentLayout: 'list',
            attachments: [attachment]
        }
    };
}

// 'Share Message' command - shares a forwarded message as a hero card,
// optionally including an image (demonstrates passing custom parameters).
function shareMessageCommand(action) {
    const userName = action.messagePayload?.from?.user?.displayName ?? 'unknown';

    const images = action.data.includeImage === 'true'
        ? ['https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQtB3AwMUeNoq4gUBGe6Ocj8kyh3bXa9ZbV7u1fVKQoyKFHdkqU']
        : [];

    const heroCard = CardFactory.heroCard(
        `${userName} originally sent this message:`,
        action.messagePayload.body.content,
        images
    );

    const attachmentCount = action.messagePayload.attachments?.length ?? 0;
    if (attachmentCount > 0) {
        // This sample does not forward MessagePayload attachments; that is left
        // as an exercise for the reader.
        heroCard.content.subtitle = `(${attachmentCount} Attachments not included)`;
    }

    const attachment = {
        contentType: heroCard.contentType,
        content: heroCard.content,
        preview: heroCard
    };

    return {
        composeExtension: {
            type: 'result',
            attachmentLayout: 'list',
            attachments: [attachment]
        }
    };
}

module.exports.BotActivityHandler = BotActivityHandler;

